server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,310 @@
+        logger.trace("The cluster has a stable master node");
+        return createIndicator(stableMasterStatus, summary, HealthIndicatorDetails.EMPTY, impacts, userActions);
+    }
+
+    /**
+     * Returns the health result for the case when we have NOT seen a master recently (at some point in the last 30 seconds).
+     * @param localMasterHistory The master history as seen from the local machine
+     * @param includeDetails Whether to calculate and include the details in the result
+     * @return The HealthIndicatorResult for the given localMasterHistory
+     */
+    private HealthIndicatorResult calculateWhenHaveNotSeenMasterRecently(MasterHistory localMasterHistory, boolean includeDetails) {
+        // NOTE: The logic in this method will be implemented in a future PR
+        HealthStatus stableMasterStatus = HealthStatus.RED;
+        String summary = "Placeholder summary";
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        List<UserAction> userActions = new ArrayList<>();
+        return createIndicator(
+            stableMasterStatus,
+            summary,
+            includeDetails ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
+            impacts,
+            userActions
+        );
+    }
+
+    private boolean hasSeenMasterInVeryRecentPast() {
+        if (clusterService.state().nodes().getMasterNode() != null) {
+            return true;
+        }
+        return masterHistoryService.getLocalMasterHistory().hasSeenMasterInLastNSeconds((int) VERY_RECENT_PAST.seconds());
+    }
+
+    /*
+     * If we detect that the same master has gone null 3 or more times, we ask the MasterHistoryService to fetch the master history
+     * as seen from that node so that it is ready in case a health API request comes in.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null && previousMaster != null) {
+            if (masterHistoryService.getLocalMasterHistory().hasMasterGoneNullAtLeastNTimes(ACCEPTABLE_NULL_TRANSITIONS + 1)) {
+                DiscoveryNode master = masterHistoryService.getLocalMasterHistory().getMostRecentNonNullMaster();
+                if (master != null && clusterService.localNode().equals(master) == false) {
+                    masterHistoryService.requestRemoteMasterHistory(master);
+                }
+            }
+        }
+    }
+
+    /**
+     * XContentBuilder doesn't deal well with ToXContentFragments (which is what DiscoveryNodes are). Also XContentBuilder doesn't do well
+     * with null values in lists. This object wraps the DiscoveryNode's XContent in a start and end object, and writes out nulls as empty
+     * objects.
```

Comment:
I think we can drop nulls from the XContent output. The history will alternate between null and not-null so the nulls can be inferred.

Reply:
Is that guaranteed (the alternating)? It's definitely what I've observed, but didn't know if that was always the case.

Reply:
Hmm. Actually no it looks like it's not guaranteed. The state changes when `Coordinator` calls `clusterApplier.onNewClusterState` but each submitted task reads the latest value of `Coordinator#applierState` so if a task runs super-slowly then it could theoretically miss an intermediate state.

I'm pretty sure this doesn't matter for the purposes of this API, however.

Reply:
This does take away one piece of information though right -- whether the currently-seen master is null or not? For example:
```
node1 -> null -> node1 -> null -> node1 -> null -> node1 -> null
```
(which with 4 null transitions (plus not currently having a master) would warrant a yellow response) would register the same as
```
node1 -> null -> node1 -> null -> node1 -> null -> node1
```
which would give a green response. I could have it leave trailing nulls, with the rest implied. But would that gain us much of anything?

Reply:
I think by the point we're turning this info into `XContent` we've already decided the colour of the response so I don't see that the difference matters. Also we return `current_master` separately from the history, and I'd prefer `"current_master": null` over `"current_master": {}` I think.

Reply:
I don't think that's the case (that we've already decided the color of the response by the time we turn this into XContent). We turn it into XContent on the remote master with zero intelligence behind it, and then interpret the meaning of it on the local host. It's the local host who decides whether the node has been flapping null or not. If it doesn't get all that information then it's going to make the wrong choice right?
Your point about `null` vs `{}` makes sense.

Reply:
> We turn it into XContent on the remote master

AFAICT we send it from the remote master as a `MasterHistoryAction.Response` (i.e. a `Writeable`) and not as an `XContent`. We typically don't use `XContent` for node-to-node comms.

Reply:
Oops. Good point!